File: frontend/src/types/moderation.ts
export type ModerationStatus =
  | 'pending'
  | 'in_review'
  | 'approved'
  | 'rejected'
  | 'escalated'
  | 'flagged';

export type ViolationCategory =
  | 'violence'
  | 'nudity'
  | 'drugs'
  | 'hate_symbols'
  | 'spam'
  | 'misinformation'
  | 'other';

export type ReviewAction = 'approve' | 'reject' | 'escalate';

export type PolicyAction = 'block' | 'flag' | 'allow';

export interface Violation {
  id: string;
  category: ViolationCategory;
  timestamp_seconds?: number;
  confidence: number;
  snippet?: string;
  frame_url?: string;
  details?: Record<string, unknown>;
}

export interface ModerationReport {
  violations: Violation[];
  recommended_action: ReviewAction;
  policy_id?: string;
  summary?: string;
  processed_at: string;
}

// Per-violation finding from the AI pipeline
export interface AiViolation {
  category: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  confidence: number;
  timestamp: number;
  description: string;
  frame_index?: number;
}

export interface ModerationQueueItem {
  id: string;
  video_id?: string;
  stream_id?: string;
  moderation_result_id?: string;
  status: ModerationStatus;
  priority: number;
  report?: ModerationReport; // legacy field
  reviewed_by?: string;
  reviewed_at?: string;
  review_notes?: string;
  created_at: string;
  updated_at?: string;
  // Joined from ModerationResult — AI justification
  video_title?: string;
  ai_summary?: string;
  overall_confidence?: number;
  violations?: AiViolation[];
  ai_model?: string;
}

export interface ModerationQueueResponse {
  items: ModerationQueueItem[];
  total: number;
  page: number;
  page_size: number;
}

export type ModerationResultResponse = ModerationQueueItem;

export interface SubmitReviewRequest {
  action: ReviewAction;
  notes?: string;
}

export interface PolicyRule {
  category: ViolationCategory;
  threshold: number;
  action: PolicyAction;
}

export interface Policy {
  id: string;
  name: string;
  description?: string;
  is_active: boolean;
  is_default: boolean;
  rules: PolicyRule[] | null;
  default_action: PolicyAction;
  owner_id: string;
  tenant_id?: string;
  created_at: string;
  updated_at: string;
}

export interface PolicyListResponse {
  items: Policy[];
  total: number;
}

export type PolicyResponse = Policy;

export interface CreatePolicyRequest {
  name: string;
  description?: string;
  is_active?: boolean;
  is_default?: boolean;
  rules?: PolicyRule[];
  default_action?: PolicyAction;
}

export interface ModerationQueueParams {
  page?: number;
  page_size?: number;
  status?: ModerationStatus;
  priority?: number;
}

File: backend/app/schemas/moderation.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict

from app.models.moderation import ModerationStatus, ReviewAction, ViolationCategory


class ViolationItem(BaseModel):
    category: ViolationCategory
    confidence: float
    timestamp_seconds: float | None = None
    frame_url: str | None = None
    snippet: str | None = None
    details: dict[str, Any] | None = None


class ModerationResultResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    video_id: uuid.UUID
    status: ModerationStatus
    overall_confidence: float | None
    violations: list[ViolationItem] | None = None
    summary: str | None
    ai_model: str | None
    processing_time_ms: int | None
    reviewed_by: uuid.UUID | None
    review_action: ReviewAction | None
    review_notes: str | None
    reviewed_at: str | None
    override_by: uuid.UUID | None
    override_decision: str | None
    override_at: str | None
    created_at: datetime
    updated_at: datetime


class ModerationQueueItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    video_id: uuid.UUID
    moderation_result_id: uuid.UUID | None
    status: ModerationStatus
    priority: int
    assigned_to: uuid.UUID | None
    tenant_id: str | None
    created_at: datetime

    # Joined from Video
    video_title: str | None = None

    # Joined from ModerationResult — AI justification
    ai_summary: str | None = None          # content summary + reasoning + policy triggers
    overall_confidence: float | None = None
    violations: list[dict[str, Any]] = []  # each has category, severity, description, confidence
    ai_model: str | None = None


class SubmitReviewRequest(BaseModel):
    action: ReviewAction
    notes: str | None = None


class OverrideRequest(BaseModel):
    decision: str
    reason: str | None = None


class PaginatedQueue(BaseModel):
    items: list[ModerationQueueItemResponse]
    total: int
    page: int
    page_size: int


class ClearQueueResponse(BaseModel):
    removed: int
